feat: setTransactionRetentionDuration

vsmanish1772 opened the pull request:
feat: add SetTransaction retention duration filtering

## What changes are proposed in this pull request?

This PR implements the `delta.setTransactionRetentionDuration` table property in delta-kernel-rs, which allows filtering out expired transaction identifiers based on their `lastUpdated` timestamp. This feature helps prevent unbounded growth of transaction logs and reduces checkpoint file sizes.

### Key changes:
1. **Transaction filtering during log replay**: Added retention timestamp filtering to `SetTransactionVisitor` to skip transactions older than the retention duration
2. **Checkpoint writing**: Updated `CheckpointVisitor` to exclude expired transactions when creating checkpoints
3. **Backward compatibility**: Transactions without `lastUpdated` field are preserved forever (never expire)
4. **Consistent behavior**: The same retention logic is applied in both checkpoint writing and batch transaction reading operations

### Implementation details:
- Modified `SetTransactionVisitor` to accept an optional retention timestamp
- Updated `CheckpointLogReplayProcessor` to calculate and propagate transaction retention timestamps
- Enhanced `CheckpointVisitor` to filter transactions based on retention during checkpoint creation
- Single transaction lookups (`get_app_id_version`) intentionally do filter to 
- Batch operations (future `get_all_app_id_versions`, checkpointing) do filter expired transactions

## How was this change tested?

Added comprehensive unit tests covering:

1. **Checkpoint writing tests** (`checkpoint::log_replay::tests`):
   - `test_checkpoint_visitor_txn_retention`: Verifies individual transaction filtering with various timestamp scenarios
   - `test_checkpoint_actions_iter_with_txn_retention`: Tests transaction filtering across multiple batches during checkpoint creation

2. **Transaction reading tests** (`actions::set_transaction::tests`):
   - `test_visitor_retention_with_null_last_updated`: Ensures transactions without `lastUpdated` are never filtered
   - Modified existing tests to support the new retention timestamp parameter

3. **Test scenarios covered**:
   - Transactions with `lastUpdated <= retention_timestamp` are filtered out
   - Transactions with `lastUpdated = None` are always kept (backward compatibility)
   - Boundary condition: transactions exactly at retention timestamp are filtered
   - Cross-batch behavior in checkpoint processing
   - Verification that filtered transactions are excluded from checkpoint files

All tests pass with `cargo t --all-features --all-targets`.

vsmanish1772 commented:
I think I should not pass optional retention_ts. I should set a default value as a fall back option. Thats still TBD in this pr

scovich commented on kernel/src/snapshot.rs:
```diff
@@ -363,6 +370,30 @@ impl Snapshot {
     }
+
+    /// Calculate retention timestamp for transactions
+    fn calculate_transaction_retention_timestamp(&self) -> DeltaResult<Option<i64>> {
```
This logic is almost 100% duplicated from above. Can we make it an associated function that takes `&TableProperties` as input, which both call sites can just invoke?

Also, timestamp calculations in general have lots of duplicated logic, see https://github.com/delta-io/delta-kernel-rs/issues/877.

vsmanish1772 replied:
you mean its duplicated code from `deleted_file_retention_timestamp` method in CheckpointWriter? if so, I agree. I have copied that code. Initially, I wanted to create a shared function for both the uses cases but was waiting for some reviews on it.

vsmanish1772 replied:
Agreed. I initially wanted to create a utility shared function but wanted for some opinions. Il see what I can do with 
#877

vsmanish1772 commented on kernel/src/actions/visitors.rs:
```diff
@@ -301,14 +301,18 @@ pub(crate) type SetTransactionMap = HashMap<String, SetTransaction>;
     pub(crate) application_id: Option<String>,
+    /// Minimum timestamp for transaction retention. Transactions with last_updated
+    /// older than this timestamp will be filtered out. None means no filtering.
+    retention_timestamp: Option<i64>,
```
That makes sense because that timestamp says when transactions expire. I have made changes to filtering and probably missed the doc comment. il make it clear. Thanks

scovich commented on kernel/src/utils.rs:
```diff
@@ -11,6 +14,29 @@ macro_rules! require {
+/// Returns None if set_transaction_retention_duration is not set.
+pub(crate) fn calculate_transaction_expiration_timestamp(
+    table_properties: &TableProperties,
+) -> DeltaResult<Option<i64>> {
+    match table_properties.set_transaction_retention_duration {
+        Some(duration) => {
+            let now = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .map_err(|e| Error::generic(format!("Failed to get current time: {}", e)))?;
```
This mix of result and option is tricky... does this help?
```rust
table_properties
    .set_transaction_retention_duration
    .map(|duration| {
        let now_ms = i64::try_from(...).map_err(|_| ...)?;
          ...
        Ok(now_ms - expiration_ms)
    })
    .transpose()
```

vsmanish1772 replied:
Ah I see. using map().transpose()  is much cleaner for handling the `Option<Result>` pattern. I wasn't familiar with the `transpose()` method, but now I see how it elegantly flips `Option<Result<T>>` to `Result<Option<T>>`, which is exactly what we need here.

vsmanish1772 commented on kernel/src/checkpoint/log_replay.rs:
```diff
@@ -361,12 +372,28 @@ impl CheckpointVisitor<'_> {
+        // Check retention if last_updated is present
+        if let Some(retention_ts) = self.txn_expiration_timestamp {
+            if let Some(last_updated) = getter[12].get_opt(i, "txn.lastUpdated")? {
+                let last_updated: i64 = last_updated;
+                if last_updated <= retention_ts {
```
Umm, coming from python background I would easily understand the code `if last_updated <= retention_ts`  at least from code readability.

zachschuermann commented on kernel/src/actions/set_transaction.rs:
```diff
@@ -91,8 +99,10 @@ mod tests {
+    use crate::utils::test_utils::parse_json_batch;
     use crate::Table;
 
+    use arrow_55::array::StringArray;
```
```suggestion
    use crate::arrow::array::StringArray;
```

zachschuermann commented on kernel/tests/data/app-txn-with-last-updated/part-00001-2e4c9beb-884e-4749-aaee-02a1a4f05750-c000.snappy.parquet:
do we need any data files in the test table?

vsmanish1772 commented:
@zachschuermann do you anything from my side to get this approved ?

zachschuermann commented:
> @zachschuermann do you anything from my side to get this approved ?

let's add that `#[allow(clippy::too_many_arguments)]` to get tests all green then i'll take another pass as soon as I can! (currently at a conference with limited availability)

scovich commented on kernel/src/checkpoint/log_replay.rs:
```diff
@@ -205,9 +212,12 @@ pub(crate) struct CheckpointVisitor<'seen> {
 }
 
 #[allow(unused)]
+#[allow(clippy::too_many_arguments)]
```
Probably better to add it to the `new` method that actually has the problem? Or was clippy flagging multiple methods?

vsmanish1772 commented:
> > @zachschuermann do you anything from my side to get this approved ?
> 
> let's add that `#[allow(clippy::too_many_arguments)]` to get tests all green then i'll take another pass as soon as I can! (currently at a conference with limited availability)

I have addressed the comment and is ready for another pass @zachschuermann

zachschuermann commented on kernel/src/checkpoint/log_replay.rs:
```diff
@@ -361,12 +373,27 @@ impl CheckpointVisitor<'_> {
+        // Check retention if last_updated is present
+        if let Some(retention_ts) = self.txn_expiration_timestamp {
+            if let Some(last_updated) = getter[12].get_opt(i, "txn.lastUpdated")? {
+                if i64::le(&last_updated, &retention_ts) {
```
nit: why not `last_update <= retention_ts`? casting issue?

vsmanish1772 replied:
I have applied as suggestion from Ryan not any particular reason. initially I have implemented as you mentioned but I thought Ryan's suggestion was Rust way of doing things (I'm completely from python background)

zachschuermann reviewed (approved):
LGTM just a couple quick things! and can we add docs on this behavior to our overall `get_app_id_version` API - perhaps just a little doc comment that says we filter based on the table property + lastUpdated

sorry for the delay on the review, thanks for the contribution!!
